File: nextjs/lib/copilot-agent.js
/**
 * GitHub Copilot SDK Agent 后端
 * 
 * 这是 Copilot SDK 的官方集成示例
 * 它使用 Copilot 内置的 Agentic Core 来自动编排 Skills
 * 
 * 工作流：
 * 1. 前端发送用户请求 + 文件路径
 * 2. 后端初始化 CopilotClient (用 COPILOT_GITHUB_TOKEN 认证)
 * 3. 创建 Session，注册 5 个 Skills (Tools)
 * 4. Copilot Agent 自动决定何时调用哪个 Skill
 * 5. 流式推送事件到前端
 */

import { CopilotClient, SkillParameter } from "@github/copilot-sdk";
import { spawn } from "child_process";
import path from "path";
import { fileURLToPath } from "url";
import fs from "fs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const projectRoot = path.resolve(__dirname, "../../");

/**
 * 初始化 Copilot Client
 * 这里使用官方的 Copilot SDK，会自动使用环境变量中的 COPILOT_GITHUB_TOKEN
 */
function initializeCopilotClient() {
  const token = process.env.COPILOT_GITHUB_TOKEN;
  if (!token) {
    throw new Error(
      "COPILOT_GITHUB_TOKEN 未设置。请在 .env 中添加你的 token。"
    );
  }

  return new CopilotClient({
    token: token,
    // 使用官方的生产级 Agent 运行时
    agentic: true,
    // 启用流式事件
    streaming: true,
  });
}

/**
 * 执行 Python Skill
 * 
 * 这是对 Python 后端的调用包装器
 * 实际的业务逻辑（数据处理、转换等）还在 Python 中
 */
function executePythonSkill(skillName, args) {
  return new Promise((resolve, reject) => {
    const pythonPath = process.env.PYTHON_PATH || 
      path.resolve(projectRoot, "venv", "Scripts", "python");

    // 使用 JSON.stringify 正确转义 Windows 路径中的反斜杠
    const projectRootPy = JSON.stringify(projectRoot);
    
    const pythonCode = `
import sys
import json
sys.path.insert(0, ${projectRootPy})

from agent import *

try:
    if '${skillName}' == 'read_source_data':
        result = mcp_read_source_data('${args[0]}')
    elif '${skillName}' == 'retrieve_sdtm_rules':
        result = mcp_retrieve_sdtm_rules('${args[0]}', ${args[1] ? `'${args[1]}'` : 'None'})
    elif '${skillName}' == 'propose_column_mapping':
        result = mcp_propose_column_mapping('${args[0]}', '${args[1]}')
    elif '${skillName}' == 'transform_to_sdtm':
        result = mcp_transform_to_sdtm('${args[0]}', '${args[1]}')
    elif '${skillName}' == 'validate_sdtm':
        result = mcp_validate_sdtm('${args[0]}', '${args[1]}')
    else:
        result = {'success': False, 'error': 'Unknown skill'}
    print(json.dumps(result))
except Exception as e:
    print(json.dumps({'success': False, 'error': str(e)}))
`;

    const proc = spawn(pythonPath, ["-c", pythonCode], {
      cwd: projectRoot,
      stdio: ["ignore", "pipe", "pipe"],
      env: {
        ...process.env,
        PYTHONPATH: projectRoot,
      },
    });

    let stdout = "";
    let stderr = "";

    const timeout = setTimeout(() => {
      proc.kill();
      reject(new Error(`Skill execution timeout: ${skillName}`));
    }, 10 * 60 * 1000); // 10 分钟超时

    proc.stdout.on("data", (d) => (stdout += d.toString()));
    proc.stderr.on("data", (d) => (stderr += d.toString()));

    proc.on("close", (code) => {
      clearTimeout(timeout);
      if (code !== 0) {
        return reject(
          new Error(`Skill failed (${skillName}): ${stderr || stdout}`)
        );
      }
      try {
        resolve(JSON.parse(stdout));
      } catch (e) {
        reject(
          new Error(
            `Failed to parse skill output: ${e.message}\nOutput: ${stdout}`
          )
        );
      }
    });
  });
}

/**
 * 定义 SDTM 转换的 Skills (Tools)
 * 这些会被注册到 Copilot Agent
 */
function defineSkills() {
  return [
    {
      name: "read_source_data",
      description:
        "读取并分析源数据文件（Excel/CSV），返回列信息、数据类型、缺失率、样本值",
      parameters: [
        {
          name: "source_path",
          description: "源数据文件路径（相对于项目根目录）",
          type: "string",
          required: true,
        },
      ],
      execute: async (params) => {
        return await executePythonSkill("read_source_data", [params.source_path]);
      },
    },
    {
      name: "retrieve_sdtm_rules",
      description:
        "从 RAG 检索 SDTMIG 规范，查询与特定域或变量相关的规则",
      parameters: [
        {
          name: "query",
          description: "查询内容（如：'AE domain requirements' 或 'AESEQ'）",
          type: "string",
          required: true,
        },
        {
          name: "domain",
          description: "SDTM 域名（可选，如 'AE', 'CM', 'LB'）",
          type: "string",
          required: false,
        },
      ],
      execute: async (params) => {
        return await executePythonSkill("retrieve_sdtm_rules", [
          params.query,
          params.domain,
        ]);
      },
    },
    {
      name: "propose_column_mapping",
      description: "提议源数据列与 SDTM 标准列的对应关系",
      parameters: [
        {
          name: "source_path",
          description: "源数据文件路径",
          type: "string",
          required: true,
        },
        {
          name: "domain",
          description: "SDTM 域名（如 'AE', 'CM'）",
          type: "string",
          required: true,
        },
      ],
      execute: async (params) => {
        return await executePythonSkill("propose_column_mapping", [
          params.source_path,
          params.domain,
        ]);
      },
    },
    {
      name: "transform_to_sdtm",
      description: "根据映射关系执行 SDTM 转换",
      parameters: [
        {
          name: "source_path",
          description: "源数据文件路径",
          type: "string",
          required: true,
        },
        {
          name: "domain",
          description: "SDTM 域名",
          type: "string",
          required: true,
        },
      ],
      execute: async (params) => {
        return await executePythonSkill("transform_to_sdtm", [
          params.source_path,
          params.domain,
        ]);
      },
    },
    {
      name: "validate_sdtm",
      description: "验证转换后的 SDTM 数据是否符合规范",
      parameters: [
        {
          name: "sdtm_path",
          description: "SDTM 文件路径（xlsx）",
          type: "string",
          required: true,
        },
        {
          name: "domain",
          description: "SDTM 域名",
          type: "string",
          required: true,
        },
      ],
      execute: async (params) => {
        return await executePythonSkill("validate_sdtm", [
          params.sdtm_path,
          params.domain,
        ]);
      },
    },
  ];
}

/**
 * 创建 SDTM 转换用的系统提示词
 * 这会被发送给 Copilot Agent 来指导其行为
 */
function getSystemPrompt() {
  return `你是一名高级临床数据 AI 程序员，擅长多域 SDTM 转换。你必须遵循 SDTMIG v3.3 规范。

## 支持的医学数据域（SDTM Domains）
- **AE（不良事件）**：用于记录研究期间发生的不良事件。
- **CM（伴随用药）**：用于记录研究期间的伴随用药信息。
- **LB（实验室检查）**：用于记录实验室检测结果。
- **VS（生命体征）**：用于记录血压、心率等生命体征数据。
- **DM（人口统计学）**：用于记录受试者的人口学特征（年龄、性别、种族等）。

## 工作流（必须按顺序）

1. **识别目标域**：如果用户未明确指定，根据文件内容自动检测（文件名、列名、样本值）。
2. **调用 read_source_data skill**：分析源数据的列名、数据类型、缺失率、样本值。
3. **调用 retrieve_sdtm_rules skill**：查询对应域的 SDTMIG 规范要求和关键变量定义。
4. **调用 propose_column_mapping skill**：提议源列与 SDTM 标准列的对应关系，标记缺失的必需字段。
5. **获取用户确认**（如有歧义）：显示映射预览，请用户确认或提供派生规则。
6. **调用 transform_to_sdtm skill**：执行转换、标准化、生成 SDTM 表。
7. **调用 validate_sdtm skill**：进行质量检查（必需变量完整性、缺失率、序列号连续性）。
8. **生成最终报告**：包括映射关系 JSON、数据质量问题、修复建议。

## 必须规避的错误

- **STUDYID/USUBJID**：不能盲目选择；必须先看数据取值，再选择更合理的列。
- **日期格式**：必须转为 ISO 8601 格式（YYYY-MM-DD 或 YYYY-MM-DDTHH:MM:SS）。
- **受控术语映射**：需按 SDTM 标准列表或本地编码表映射，不得低阈值强配。
- **必需变量缺失**：若映射中缺少任何必填变量（Req），必须追问用户或标记为错误。
- **序列号**：如果域有 key_sequence_var（如 AESEQ、CMSEQ），必须按记录顺序递增分配（1,2,3,...）。
- **数据丢失**：禁止默认过滤记录；除非用户明确要求。

## 关键原则

1. **规范优先**：始终遵循 SDTMIG v3.3，不得因方便而破坏规范。
2. **完整性优先**：所有必需变量必须存在，不能跳过。
3. **可追踪性**：导出映射和报告，便于审计和复查。
4. **用户参与**：在关键决策点追问用户，不要假设。
5. **智能决策**：根据实际数据情况动态调整转换策略。

你现在拥有一组 Skills（Tools）来完成任务。根据用户的需求自主决策：
- 什么时候调用 read_source_data
- 什么时候调用 retrieve_sdtm_rules
- 什么时候调用 propose_column_mapping
- 什么时候调用 transform_to_sdtm
- 什么时候调用 validate_sdtm

不要等待指示，自主完成整个 SDTM 转换任务。`;
}

export { initializeCopilotClient, defineSkills, getSystemPrompt, executePythonSkill };

File: core/copilot_agent.py
#!/usr/bin/env python
"""
Copilot SDK Integration - SDTM Multi-Domain Agent
该脚本可被 Copilot SDK 的 Node.js 后端直接调用
"""
import os
import sys
import json
import tempfile
from pathlib import Path

# 假设工作目录是 session dir
from core.sdtm_converter import (
    SDTMTransformer, 
    ValidationReport, 
    process_sdtm_conversion,
    DOMAIN_METADATA
)
from core.rag_retriever import get_retriever, format_retrieved


class SDTMCopilotAgent:
    """Copilot SDK Agent 接口"""
    
    SYSTEM_PROMPT = """
你是一名高级临床数据 AI 程序员，擅长多域 SDTM 转换。你必须遵循 SDTMIG v3.3 规范。

## 支持的医学数据域（SDTM Domains）
- **AE（不良事件）**：用于记录研究期间发生的不良事件。
- **CM（伴随用药）**：用于记录研究期间的伴随用药信息。
- **LB（实验室检查）**：用于记录实验室检测结果。
- **VS（生命体征）**：用于记录血压、心率等生命体征数据。
- **DM（人口统计学）**：用于记录受试者的人口学特征（年龄、性别、种族等）。

## 工作流（必须按顺序）

1. **识别目标域**：如果用户未明确指定，根据文件内容自动检测（文件名、列名、样本值）。
2. **调用 read_source_data skill**：分析源数据的列名、数据类型、缺失率、样本值。
3. **调用 retrieve_sdtm_rules skill**：查询对应域的 SDTMIG 规范要求和关键变量定义。
4. **调用 propose_column_mapping skill**：提议源列与 SDTM 标准列的对应关系，标记缺失的必需字段。
5. **获取用户确认**（如有歧义）：显示映射预览，请用户确认或提供派生规则。
6. **调用 transform_to_sdtm skill**：执行转换、标准化、生成 SDTM 表。
7. **调用 validate_sdtm skill**：进行质量检查（必需变量完整性、缺失率、序列号连续性）。
8. **生成最终报告**：包括映射关系 JSON、数据质量问题、修复建议。

## 必须规避的错误

- **STUDYID/USUBJID**：不能盲目选择；必须先看数据取值，再选择更合理的列。
- **日期格式**：必须转为 ISO 8601 格式（YYYY-MM-DD 或 YYYY-MM-DDTHH:MM:SS）。
- **受控术语映射**：需按 SDTM 标准列表或本地编码表映射，不得低阈值强配。
- **必需变量缺失**：若映射中缺少任何必填变量（Req），必须追问用户或标记为错误。
- **序列号**：如果域有 key_sequence_var（如 AESEQ、CMSEQ），必须按记录顺序递增分配（1,2,3,...）。
- **数据丢失**：禁止默认过滤记录；除非用户明确要求。

## 输出要求

- **SDTM_{domain}.xlsx**：转换后的数据表，包含所有必需和期望变量。
- **mapping_{domain}.json**：源列 → SDTM 列的映射关系文档（便于人工审核）。
- **report_{domain}.json**：数据质量报告（问题列表、严重级别、摘要统计）。

## 交互指南

- 如果用户只给文件路径，不给域名：询问用户这是哪个 SDTM 域的数据。
- 如果检测到映射缺陷：标记问题，请用户补充信息或确认派生规则。
- 如果数据质量差（>50% 缺失）：警告用户，确认是否继续。
- 最终完成：总结输出文件位置、映射关系、问题清单。

## 关键原则

1. **规范优先**：始终遵循 SDTMIG v3.3，不得因方便而破坏规范。
2. **完整性优先**：所有必需变量必须存在，不能跳过。
3. **可追踪性**：导出映射和报告，便于审计和复查。
4. **用户参与**：在关键决策点追问用户，不要假设。
""".strip()
    
    def __init__(self, session_dir: str):
        self.session_dir = session_dir
        # 尝试从多个位置找到 rag_store
        possible_rag_stores = [
            os.path.join(os.path.dirname(__file__), "rag_store"),
            os.path.join(os.path.dirname(os.path.dirname(__file__)), "sdtm-multi-domain-agent", "rag_store"),
            "rag_store",
        ]
        
        self.rag_store = None
        for path in possible_rag_stores:
            if os.path.exists(os.path.join(path, "sdtmig.faiss")):
                self.rag_store = os.path.abspath(path)
                break
        
        # 如果没找到，用默认路径（会在使用时报错）
        if not self.rag_store:
            self.rag_store = os.path.join(os.path.dirname(__file__), "rag_store")
        
    def list_supported_domains(self) -> dict:
        """列出支持的域"""
        return {
            "domains": list(DOMAIN_METADATA.keys()),
            "descriptions": {
                domain: DOMAIN_METADATA[domain]["description"]
                for domain in DOMAIN_METADATA.keys()
            }
        }
    
    def read_and_analyze_source(self, source_path: str) -> dict:
        """读取并分析源数据"""
        import pandas as pd
        
        if not os.path.exists(source_path):
            return {"error": f"Source file not found: {source_path}"}
        
        try:
            if source_path.endswith(('.xlsx', '.xls')):
                df = pd.read_excel(source_path)
            elif source_path.endswith('.csv'):
                df = pd.read_csv(source_path)
            else:
                return {"error": f"Unsupported file format: {source_path}"}
            
            transformer = SDTMTransformer("AE")  # 暂时用 AE 只是为了 infer_source_schema
            schema = transformer.infer_source_schema(df)
            
            return {
                "success": True,
                "shape": list(df.shape),
                "columns": schema["columns"],
                "dtypes": schema["dtypes"],
                "missing_rate": schema.get("missing_rates", {}),
                "missing_rates": schema.get("missing_rates", {}),
                "sample_values": schema["sample_values"],
            }
        except Exception as e:
            return {"error": str(e)}
    
    def retrieve_sdtm_specification(self, query: str, domain: str = None) -> dict:
        """检索 SDTM 规范"""
        try:
            if not self.rag_store or not os.path.exists(self.rag_store):
                return {
                    "success": False,
                    "error": f"RAG store not found at {self.rag_store}",
                    "query": query,
                    "domain": domain
                }
            
            retriever = get_retriever(self.rag_store)
            
            # 如果指定了域，增强查询
            if domain:
                enhanced_query = f"[{domain}] {query}"
            else:
                enhanced_query = query
            
            chunks = retriever.retrieve(enhanced_query, top_k=5)
            
            return {
                "success": True,
                "query": query,
                "domain": domain,
                "chunks_count": len(chunks),
                "results": [
                    {
                        "page": c.page,
                        "score": round(float(c.score), 4),
                        "text": c.text,
                    }
                    for c in chunks
                ],
                "formatted": format_retrieved(chunks),
            }
        except Exception as e:
            import traceback
            return {
                "success": False,
                "error": str(e),
                "traceback": traceback.format_exc(),
                "query": query,
                "domain": domain
            }
            return {"error": str(e)}
    
    def propose_mapping_strategy(self, source_path: str, domain: str) -> dict:
        """提议映射策略（包含 RAG 规范参考）"""
        import pandas as pd
        
        if domain not in DOMAIN_METADATA:
            return {"error": f"Unsupported domain: {domain}"}
        
        if not os.path.exists(source_path):
            return {"error": f"Source file not found: {source_path}"}
        
        try:
            if source_path.endswith(('.xlsx', '.xls')):
                df = pd.read_excel(source_path)
            elif source_path.endswith('.csv'):
                df = pd.read_csv(source_path)
            else:
                return {"error": f"Unsupported file format: {source_path}"}
            
            # 先检索该域的规范
            rag_result = self.retrieve_sdtm_specification(
                f"Domain {domain} required variables and standards",
                domain=domain
            )
            rag_context = rag_result.get("formatted", "") if rag_result.get("success") else ""
            
            transformer = SDTMTransformer(domain)
            schema = transformer.infer_source_schema(df)
            mapping = transformer.propose_mapping(schema, rag_context=rag_context)
            
            return {
                "success": True,
                "domain": domain,
                "proposed_mapping": mapping,
                "required_vars": DOMAIN_METADATA[domain]["required_vars"],
                "expected_vars": DOMAIN_METADATA[domain]["expected_vars"],
                "unmapped_columns": [
                    col for col in schema["columns"]
                    if col not in mapping
                ],
                "rag_context_available": bool(rag_context),
                "issues": transformer.issues,
                "guidance": self._generate_mapping_guidance(transformer.issues),
            }
        except Exception as e:
            import traceback
            return {
                "success": False,
                "error": str(e),
                "traceback": traceback.format_exc()
            }
    
    def execute_conversion(self, source_path: str, domain: str, mapping: dict = None) -> dict:
        """执行 SDTM 转换"""
        import pandas as pd
        
        if domain not in DOMAIN_METADATA:
            return {"success": False, "error": f"Unsupported domain: {domain}"}
        
        if not os.path.exists(source_path):
            return {"success": False, "error": f"Source file not found: {source_path}"}
        
        try:
            if source_path.endswith(('.xlsx', '.xls')):
                df_source = pd.read_excel(source_path)
            elif source_path.endswith('.csv'):
                df_source = pd.read_csv(source_path)
            else:
                return {"success": False, "error": f"Unsupported file format: {source_path}"}
            
            transformer = SDTMTransformer(domain)
            
            # 如果提供了映射，使用；否则自动提议
            if mapping is None:
                schema = transformer.infer_source_schema(df_source)
                mapping = transformer.propose_mapping(schema)
            else:
                # 确保 mapping 是字典
                if isinstance(mapping, str):
                    mapping = json.loads(mapping)
                transformer.mapping = mapping
            
            # 标准化
            df_std, std_issues = transformer.standardize_data(df_source)
            
            # 转换
            df_sdtm, trans_issues = transformer.apply_mapping(df_std)
            
            # 验证
            validator = ValidationReport(domain, df_sdtm)
            validation_report = validator.validate()
            
            # 确保 session_dir 存在
            os.makedirs(self.session_dir, exist_ok=True)
            
            # 输出文件
            sdtm_file = os.path.join(self.session_dir, f"SDTM_{domain}.xlsx")
            df_sdtm.to_excel(sdtm_file, index=False, engine='openpyxl')
            
            mapping_file = os.path.join(self.session_dir, f"mapping_{domain}.json")
            with open(mapping_file, 'w', encoding='utf-8') as f:
                json.dump(mapping, f, indent=2, ensure_ascii=False)
            
            report_file = os.path.join(self.session_dir, f"report_{domain}.json")
            with open(report_file, 'w', encoding='utf-8') as f:
                json.dump(validation_report, f, indent=2, ensure_ascii=False)
            
            # 确保所有返回值都是可序列化的
            return {
                "success": True,
                "domain": domain,
                "output_shape": list(df_sdtm.shape),
                "sdtm_file": str(sdtm_file),
                "mapping_file": str(mapping_file),
                "report_file": str(report_file),
                "rows_processed": int(len(df_sdtm)),
                "columns_created": int(len(df_sdtm.columns)),
                "validation_issue_count": len(validation_report.get("issues", [])),
                "validation_errors": len([i for i in validation_report.get("issues", []) if i.get("severity") == "error"]),
                "validation_warnings": len([i for i in validation_report.get("issues", []) if i.get("severity") == "warning"]),
            }
        
        except Exception as e:
            import traceback
            return {
                "success": False,
                "error": str(e),
                "traceback": traceback.format_exc(),
                "domain": domain
            }
    
    def _generate_mapping_guidance(self, issues: list) -> str:
        """生成映射指导"""
        if not issues:
            return "Mapping looks good! Proceed with conversion."
        
        guidance = "Mapping Issues Found:\n"
        for issue in issues:
            if issue["type"] == "missing_required_var":
                guidance += f"⚠️  Required variable '{issue['var']}' is missing. " \
                           f"Please specify which source column maps to it, or provide a derivation rule.\n"
        
        return guidance


def main():
    """CLI 入口（用于测试或被外部调用）"""
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: copilot_agent.py <command> [args...]"}, ensure_ascii=False))
        sys.exit(1)
    
    # 使用当前工作目录作为 session dir
    agent = SDTMCopilotAgent(os.getcwd())
    
    command = sys.argv[1]
    
    try:
        if command == "list_domains":
            result = agent.list_supported_domains()
        elif command == "read":
            result = agent.read_and_analyze_source(sys.argv[2])
        elif command == "retrieve":
            domain = sys.argv[3] if len(sys.argv) > 3 else None
            result = agent.retrieve_sdtm_specification(sys.argv[2], domain)
        elif command == "propose":
            result = agent.propose_mapping_strategy(sys.argv[2], sys.argv[3])
        elif command == "convert":
            mapping = json.loads(sys.argv[4]) if len(sys.argv) > 4 else None
            result = agent.execute_conversion(sys.argv[2], sys.argv[3], mapping)
        else:
            result = {"error": f"Unknown command: {command}"}
        
        print(json.dumps(result, ensure_ascii=False, indent=2))
    
    except Exception as e:
        import traceback
        print(json.dumps({
            "error": str(e),
            "traceback": traceback.format_exc()
        }, ensure_ascii=False))
        sys.exit(1)


if __name__ == "__main__":
    main()
